targets/TARGET_Silicon_Labs/TARGET_EFM32/trng/sl_trng.c

```diff
@@ -0,0 +1,274 @@
+  {0xE1BCC06B, 0x9199452A, 0x1A7434E1, 0x25199E7F,
+   0x578A2DAE, 0x9CAC031E, 0xAC6FB79E, 0x518EAF45,
+   0x461CC830, 0x11E45CA3, 0x19C1FBE5, 0xEF520A1A,
+   0x45249FF6, 0x179B4FDF, 0x7B412BAD, 0x10376CE6};
+
+#define TEST_VECTOR_CONDITIONING_OUTPUT_SIZE (4)
+static const uint32_t
+test_vector_conditioning_output[TEST_VECTOR_CONDITIONING_OUTPUT_SIZE] =
+  {0xA1CAF13F, 0x09AC1F68, 0x30CA0E12, 0xA7E18675};
+
+#define TRNG_STARTUP_TEST_WAIT_RETRY  (10000)
+
+typedef struct {
+  TRNG_TypeDef *instance;
+  CMU_Clock_TypeDef clock;
+} sl_trng_device_t;
+
+static const sl_trng_device_t sl_trng_devices[TRNG_COUNT] = 
+{
+#if defined(TRNG0)
+{
+  TRNG0,
+  cmuClock_TRNG0
+},
+#endif
+};
+
+static CMU_Clock_TypeDef sl_trng_get_clock( TRNG_TypeDef *device )
+{
+  for(int i = 0; i < TRNG_COUNT; i++) {
+    if(sl_trng_devices[i].instance == device) {
+      return sl_trng_devices[i].clock;
+    }
+  }
+  return cmuClock_TRNG0;
+}
+
+void sl_trng_init( TRNG_TypeDef *device )
+{
+    int i;
+
+    /* Enable the TRNG's clock. */
+    CMU_ClockEnable( sl_trng_get_clock(device), true );
+
+    device->CONTROL =
+        TRNG_CONTROL_ENABLE |
+        TRNG_CONTROL_REPCOUNTIEN |
+        TRNG_CONTROL_APT64IEN |
+        TRNG_CONTROL_APT4096IEN |
+        TRNG_CONTROL_PREIEN |
+        TRNG_CONTROL_ALMIEN;
+
+    /* Apply software reset */
+    sl_trng_soft_reset(device);
+
+    /* Wait for TRNG to complete startup tests and start filling the FIFO. */
+    for (i=0; (device->FIFOLEVEL == 0) && (i<TRNG_STARTUP_TEST_WAIT_RETRY); i++);
+    EFM_ASSERT(i<TRNG_STARTUP_TEST_WAIT_RETRY);
+}
+
+void sl_trng_free( TRNG_TypeDef *device )
+{
+    /* Disable TRNG. */
+    device->CONTROL = 0;
+
+    /* Disable the TRNG clock. */
+    CMU_ClockEnable( sl_trng_get_clock(device), false );
+}
+
+void sl_trng_soft_reset( TRNG_TypeDef *device )
+{
+    uint32_t ctrl = device->CONTROL;
+
+    ctrl |= TRNG_CONTROL_SOFTRESET;
+
+    device->CONTROL = ctrl;
+
+    ctrl &= ~TRNG_CONTROL_SOFTRESET;
+    device->CONTROL = ctrl;
+}
+
+static void sl_trng_clear_fifo( TRNG_TypeDef *device )
+{
+  volatile uint32_t val32;
+
+  /* Empty FIFO */
+  while ( device->FIFOLEVEL )
+  {
+      val32 = device->FIFO;
+      (void)val32;
+  }
+}
+
+int sl_trng_set_key( TRNG_TypeDef *device, const unsigned char *key )
+{
+    uint32_t *_key = (uint32_t*) key;
+
+    sl_trng_clear_fifo(device);
+
+    /* Program key in KEY registers of the TRNG. */
+    device->KEY0 = *_key++;
+    device->KEY1 = *_key++;
+    device->KEY2 = *_key++;
+    device->KEY3 = *_key++;
+
+    return 0;
+}
+
+static int sl_trng_check_status( TRNG_TypeDef *device )
+{
+    uint32_t status = device->STATUS;
+
+    if ( (status & (TRNG_STATUS_PREIF
+                    | TRNG_STATUS_REPCOUNTIF
+                    | TRNG_STATUS_APT64IF
+                    | TRNG_STATUS_APT4096IF
+                    | TRNG_STATUS_ALMIF)) == 0 )
+    {
+        /* No errors */
+        return 0;
+    }
+
+    if ( status & TRNG_STATUS_PREIF )
```

Comment:
If the preliminary noise alarm is raised, other alarms are ignored. Is this desirable? Or is the point moot because the alarms are mutually exclusive?